File: js/lib/UniversalId.Builder.ts
import { Codebooks, Gmod, UniversalId } from ".";
import { ImoNumber } from "./ImoNumber";
import { LocalIdParsingErrorBuilder } from "./internal/LocalIdParsingErrorBuilder";
import { LocalIdBuilder } from "./LocalId.Builder";
import { Locations } from "./Location";
import { ParsingState } from "./types/LocalId";
import { UniversalIdParser } from "./UniversalId.Parsing";
import { VisVersion } from "./VisVersion";

export class UniversalIdBuilder {
    public static readonly namingEntity = "data.dnv.com";
    private _localId?: LocalIdBuilder;
    public imoNumber?: ImoNumber;

    public get localId() {
        return this._localId;
    }

    public isValid(locations: Locations): boolean {
        return !!this.imoNumber?.isValid && !!this.localId?.isValid(locations);
    }

    public static create(visVersion: VisVersion) {
        return new UniversalIdBuilder().withLocalId(
            LocalIdBuilder.create(visVersion)
        );
    }

    public equals(other?: UniversalIdBuilder): boolean {
        if (!other) return false;
        return (
            this.imoNumber === other.imoNumber &&
            !!this.localId?.equals(other.localId)
        );
    }

    public clone() {
        return Object.assign(new UniversalIdBuilder(), this);
    }

    public validate(
        locations: Locations,
        errorBuilder = new LocalIdParsingErrorBuilder()
    ): LocalIdParsingErrorBuilder {
        if (!this.imoNumber?.isValid) {
            errorBuilder.push({
                type: ParsingState.EmptyState,
                message: "Missing or invalid IMO number",
            });
        }
        this._localId?.validate(locations, errorBuilder);
        return errorBuilder;
    }

    public static parse(
        universalId: string,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return UniversalIdParser.parse(
            universalId,
            gmod,
            codebooks,
            locations,
            errorBuilder
        );
    }

    public static async parseAsync(
        universalIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return UniversalIdParser.parseAsync(universalIdString, errorBuilder);
    }

    public static tryParse(
        universalId: string,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ): UniversalIdBuilder | undefined {
        return UniversalIdParser.tryParse(
            universalId,
            gmod,
            codebooks,
            locations,
            errorBuilder
        );
    }

    public static async tryParseAsync(
        universalIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return UniversalIdParser.tryParseAsync(universalIdString, errorBuilder);
    }

    public build(): UniversalId {
        return new UniversalId(this);
    }

    public toString() {
        const builder: string[] = [];

        if (!this.imoNumber)
            throw new Error("Invalid Universal Id state: Missing IMO Number");
        if (!this.localId)
            throw new Error("Invalid Universal Id state: Missing LocalId");

        builder.push(UniversalIdBuilder.namingEntity);
        builder.push("/");
        builder.push(this.imoNumber.toString());
        this.localId.toString(builder);

        return builder.join("");
    }

    public withLocalId(localId?: LocalIdBuilder) {
        return this.with((s) => (s._localId = localId));
    }

    public withImoNumber(imoNumber?: ImoNumber) {
        return this.with((s) => (s.imoNumber = imoNumber));
    }

    public with(u: (state: UniversalIdBuilder) => void): UniversalIdBuilder {
        const n = this.clone();
        u && u(n);
        return n;
    }
}

File: cpp/src/dnv/vista/sdk/UniversalIdBuilder.cpp
#include "pch.h"

#include "dnv/vista/sdk/UniversalIdBuilder.h"

#include "dnv/vista/sdk/ImoNumber.h"
#include "dnv/vista/sdk/LocalIdBuilder.h"
#include "dnv/vista/sdk/LocalIdParsingErrorBuilder.h"
#include "dnv/vista/sdk/ParsingErrors.h"
#include "dnv/vista/sdk/UniversalId.h"
#include "dnv/vista/sdk/VISVersion.h"

namespace dnv::vista::sdk
{
	//=====================================================================
	// Constants
	//=====================================================================

	const std::string UniversalIdBuilder::namingEntity = "data.dnv.com";

	//=====================================================================
	// UniversalIdBuilder class
	//=====================================================================

	//----------------------------------------------
	// Assignment operators
	//----------------------------------------------

	UniversalIdBuilder& UniversalIdBuilder::operator=( UniversalIdBuilder&& other ) noexcept
	{
		if ( this != &other )
		{
			m_imoNumber = std::move( other.m_imoNumber );
			m_localIdBuilder = std::move( other.m_localIdBuilder );
		}

		return *this;
	}

	//----------------------------------------------
	// Operators
	//----------------------------------------------

	bool UniversalIdBuilder::operator==( const UniversalIdBuilder& other ) const
	{
		return equals( other );
	}

	bool UniversalIdBuilder::operator!=( const UniversalIdBuilder& other ) const
	{
		return !equals( other );
	}

	bool UniversalIdBuilder::equals( const UniversalIdBuilder& other ) const
	{
		return m_imoNumber == other.m_imoNumber && m_localIdBuilder == other.m_localIdBuilder;
	}

	//----------------------------------------------
	// Accessors
	//----------------------------------------------

	const std::optional<ImoNumber>& UniversalIdBuilder::imoNumber() const noexcept
	{
		return m_imoNumber;
	}

	const std::optional<LocalIdBuilder>& UniversalIdBuilder::localId() const noexcept
	{
		return m_localIdBuilder;
	}

	size_t UniversalIdBuilder::hashCode() const noexcept
	{
		size_t hash = 0;

		if ( m_imoNumber.has_value() )
			hash ^= m_imoNumber->hashCode() + 0x9e3779b9 + ( hash << 6 ) + ( hash >> 2 );

		if ( m_localIdBuilder.has_value() )
			hash ^= m_localIdBuilder->hashCode() + 0x9e3779b9 + ( hash << 6 ) + ( hash >> 2 );

		return hash;
	}

	//----------------------------------------------
	// State inspection methods
	//----------------------------------------------

	bool UniversalIdBuilder::isValid() const noexcept
	{
		return m_imoNumber.has_value() && m_localIdBuilder.has_value() && m_localIdBuilder->isValid();
	}

	//----------------------------------------------
	// String conversion
	//----------------------------------------------

	std::string UniversalIdBuilder::toString() const
	{
		if ( !m_imoNumber.has_value() )
		{
			throw std::invalid_argument( "Invalid Universal Id state: Missing IMO Number" );
		}

		if ( !m_localIdBuilder.has_value() )
		{
			throw std::invalid_argument( "Invalid Universal Id state: Missing LocalId" );
		}

		std::ostringstream builder;

		builder << namingEntity;
		builder << "/";
		builder << m_imoNumber->toString();

		builder << m_localIdBuilder->toString();

		return builder.str();
	}

	//----------------------------------------------
	// Static factory methods
	//----------------------------------------------

	UniversalIdBuilder UniversalIdBuilder::create( VisVersion version )
	{
		return UniversalIdBuilder().withLocalId( LocalIdBuilder::create( version ) );
	}

	//----------------------------------------------
	// Build methods (Immutable fluent interface)
	//----------------------------------------------

	//----------------------------
	// Build
	//----------------------------

	UniversalId UniversalIdBuilder::build() const
	{
		return UniversalId( *this );
	}

	//----------------------------------------------
	// Local id
	//----------------------------------------------

	UniversalIdBuilder UniversalIdBuilder::withLocalId( const LocalIdBuilder& localId ) const
	{
		bool succeeded;
		auto result = tryWithLocalId( localId, succeeded );
		if ( !succeeded )
		{
			throw std::invalid_argument( "withLocalId" );
		}
		return result;
	}

	UniversalIdBuilder UniversalIdBuilder::tryWithLocalId( const std::optional<LocalIdBuilder>& localId ) const
	{
		bool dummy;
		return tryWithLocalId( localId, dummy );
	}

	UniversalIdBuilder UniversalIdBuilder::tryWithLocalId( const std::optional<LocalIdBuilder>& localId, bool& succeeded ) const
	{
		if ( !localId.has_value() )
		{
			succeeded = false;
			return *this;
		}

		succeeded = true;

		UniversalIdBuilder result( *this );

		result.m_localIdBuilder.emplace( localId.value() );

		return result;
	}

	UniversalIdBuilder UniversalIdBuilder::withoutLocalId() const
	{
		UniversalIdBuilder result( *this );
		result.m_localIdBuilder = std::nullopt;
		return result;
	}

	//----------------------------------------------
	// IMO number
	//----------------------------------------------

	UniversalIdBuilder UniversalIdBuilder::withImoNumber( const ImoNumber& imoNumber ) const
	{
		bool succeeded;
		auto result = tryWithImoNumber( imoNumber, succeeded );
		if ( !succeeded )
		{
			throw std::invalid_argument( "imoNumber" );
		}
		return result;
	}

	UniversalIdBuilder UniversalIdBuilder::tryWithImoNumber( const std::optional<ImoNumber>& imoNumber ) const
	{
		bool dummy;
		return tryWithImoNumber( imoNumber, dummy );
	}

	UniversalIdBuilder UniversalIdBuilder::tryWithImoNumber( const std::optional<ImoNumber>& imoNumber, bool& succeeded ) const
	{
		if ( !imoNumber.has_value() )
		{
			succeeded = false;
			return *this;
		}

		succeeded = true;

		UniversalIdBuilder result( *this );
		result.m_imoNumber = imoNumber;

		return result;
	}

	UniversalIdBuilder UniversalIdBuilder::withoutImoNumber() const
	{
		UniversalIdBuilder result( *this );
		result.m_imoNumber = std::nullopt;
		return result;
	}

	//----------------------------------------------
	// Static parsing methods
	//----------------------------------------------

	UniversalIdBuilder UniversalIdBuilder::parse( std::string_view universalIdStr )
	{
		ParsingErrors errors;
		std::optional<UniversalIdBuilder> builder;
		if ( !tryParse( universalIdStr, errors, builder ) )
		{
			std::string errorMessage = "Couldn't parse universal ID from: '" + std::string( universalIdStr ) + "'. " + errors.toString();
			throw std::invalid_argument( errorMessage );
		}
		return builder.value();
	}

	bool UniversalIdBuilder::tryParse( std::string_view universalId, std::optional<UniversalIdBuilder>& universalIdBuilder )
	{
		ParsingErrors errors;
		return tryParse( universalId, errors, universalIdBuilder );
	}

	bool UniversalIdBuilder::tryParse( std::string_view universalId, ParsingErrors& errors, std::optional<UniversalIdBuilder>& universalIdBuilder )
	{
		universalIdBuilder = std::nullopt;
		auto errorBuilder = LocalIdParsingErrorBuilder::create();

		if ( universalId.empty() )
		{
			errorBuilder.addError( LocalIdParsingState::NamingRule, std::string( "Failed to find localId start segment" ) );
			errors = errorBuilder.build();
			return false;
		}

		auto localIdStartIndex = universalId.find( "/dnv-v" );
		if ( localIdStartIndex == std::string::npos )
		{
			errorBuilder.addError( LocalIdParsingState::NamingRule, std::string( "Failed to find localId start segment" ) );
			errors = errorBuilder.build();
			return false;
		}

		std::string universalIdSegment = std::string( universalId.substr( 0, localIdStartIndex ) );
		std::string localIdSegment = std::string( universalId.substr( localIdStartIndex ) );

		std::optional<ImoNumber> imoNumber = std::nullopt;

		std::optional<LocalIdBuilder> localIdBuilder = std::nullopt;
		ParsingErrors localIdErrors;
		if ( !LocalIdBuilder::tryParse( localIdSegment, localIdErrors, localIdBuilder ) )
		{
			errors = errorBuilder.build();
			return false;
		}

		std::string_view span = universalIdSegment;
		LocalIdParsingState state = LocalIdParsingState::NamingEntity;
		size_t i = 0;

		while ( state <= LocalIdParsingState::IMONumber )
		{
			if ( i >= span.length() )
				break;

			auto nextSlash = span.substr( i ).find( '/' );
			std::string_view segment = ( nextSlash == std::string_view::npos ) ? span.substr( i ) : span.substr( i, nextSlash );

			switch ( state )
			{
				case LocalIdParsingState::NamingRule:
				case LocalIdParsingState::VisVersion:
				case LocalIdParsingState::PrimaryItem:
				case LocalIdParsingState::SecondaryItem:
				case LocalIdParsingState::ItemDescription:
				case LocalIdParsingState::MetaQuantity:
				case LocalIdParsingState::MetaContent:
				case LocalIdParsingState::MetaCalculation:
				case LocalIdParsingState::MetaState:
				case LocalIdParsingState::MetaCommand:
				case LocalIdParsingState::MetaType:
				case LocalIdParsingState::MetaPosition:
				case LocalIdParsingState::MetaDetail:
				case LocalIdParsingState::EmptyState:
				case LocalIdParsingState::Formatting:
				case LocalIdParsingState::Completeness:
					break;
				case LocalIdParsingState::NamingEntity:
					if ( segment != namingEntity )
					{
						errorBuilder.addError( state, std::string( "Naming entity segment didnt match. Found: " ) + std::string( segment ) );
						break;
					}
					break;
				case LocalIdParsingState::IMONumber:
				{
					auto imoResult = ImoNumber::tryParse( segment );
					if ( !imoResult.has_value() )
					{
						errorBuilder.addError( state, std::string( "Invalid IMO number segment" ) );
						break;
					}
					else
					{
						imoNumber = imoResult;
					}
				}
				break;
			}
			state = static_cast<LocalIdParsingState>( static_cast<int>( state ) + 1 );
			i += segment.length() + 1;
		}

		auto visVersion = localIdBuilder->visVersion();
		if ( !visVersion.has_value() )
		{
			errorBuilder.addError( LocalIdParsingState::VisVersion );
			errors = errorBuilder.build();
			return false;
		}

		universalIdBuilder.emplace( create( visVersion.value() )
				.tryWithLocalId( localIdBuilder )
				.tryWithImoNumber( imoNumber ) );

		errors = errorBuilder.build();

		return true;
	}
}
